princexmlserver/converter.py

```diff
@@ -36,53 +63,86 @@ def _findbinary(self):
                 return fullname
         return None
 
-    def _run_command(self, cmd):
-        if isinstance(cmd, str):
-            cmd = cmd.split()
+    def _run_command(
+        self,
+        cmd,
+        outputpath=None,
+        script_file=None,
+        final_pass=True,
+    ):
+        # if isinstance(cmd, str):
+        #     cmd = cmd.split()
+        if final_pass:
+            cmd += [f'--output={outputpath}']
         cmdformatted = ' '.join(cmd)
-        logger.info("Running command %s" % cmdformatted)
+        logger.info(f'Running command {cmdformatted}')
         process = subprocess.Popen(
             cmd, stdout=subprocess.PIPE,
             stderr=subprocess.PIPE, close_fds=self.close_fds)
         output, error = process.communicate()
         process.stdout.close()
         process.stderr.close()
         if process.returncode != 0:
-            error = """Command
-%s
-finished with return code
-%i
-and output:
-%s
-%s""" % (
-                cmdformatted, process.returncode,
-                output.decode('utf-8'), error.decode('utf-8'))
-            logger.info(error)
-            raise Exception(error)
-        logger.info("Finished Running Command %s" % cmdformatted)
-        return output
+            self._notify_error(cmdformatted, process, output, error)
+        logger.info(f'Finished Running Command {cmdformatted}')
+        if os.path.exists(outputpath):
+            return output
+        if error:
+            if error.startswith(b'prince: error: page count greater than 1'):
+                with open(script_file, 'a+') as file:
+                    file.write('quarterInchPageHeight++;\n')
```

Comment:
Maybe add some documentation of what this variable is and what it does. I understand it's a global JavaScript variable incrementation, but as it's a little hacky it might be good to document.

Reply:
It also might be good to write this in a way that makes this still work even if the variable is not defined / used.

e.g.
```
> quarterInchPageHeight++;
Uncaught ReferenceError: quarterInchPageHeight is not defined
```

```
> (typeof quarterInchPageHeight === 'undefined') ? true : (quarterInchPageHeight++);
true
> var quarterInchPageHeight = 0;
undefined
> (typeof quarterInchPageHeight === 'undefined') ? true : (quarterInchPageHeight++);
0
> (typeof quarterInchPageHeight === 'undefined') ? true : (quarterInchPageHeight++);
1
```

Reply:
> Maybe add some documentation of what this variable is and what it does. I understand it's a global JavaScript variable incrementation, but as it's a little hacky it might be good to document.

good idea

Reply:
> It also might be good to write this in a way that makes this still work even if the variable is not defined / used.
> 
> e.g.
> 
> ```
> > quarterInchPageHeight++;
> Uncaught ReferenceError: quarterInchPageHeight is not defined
> ```
> 
> ```
> > (typeof quarterInchPageHeight === 'undefined') ? true : (quarterInchPageHeight++);
> true
> > var quarterInchPageHeight = 0;
> undefined
> > (typeof quarterInchPageHeight === 'undefined') ? true : (quarterInchPageHeight++);
> 0
> > (typeof quarterInchPageHeight === 'undefined') ? true : (quarterInchPageHeight++);
> 1
> ```

```(typeof 'quarterInchPageHeight !== 'undefined') && quarterInchPageHeight++;```
I'll through that in, but my thought was for now, it would only error if someone send through a different js file (which would likely require a slight server rewrite anyway, which could handle that), but no harm in doing it.